### PR TITLE
Organization member onboarding using the organization identity provider

### LINKED ISSUE #28273: Register users as an organization member when authenticating through the identity provider bound to an organization
This task is about onboarding organization members by federating their accounts from the identity provider associated with an organization.

The basic flow will be as follows:

* By reaching the authorization endpoint, the server will first ask for the username (no credentials). In other words, executing an identity-first login.
* Once the user provides a username and the value is an e-mail that matches the domain of an organization, the user is automatically redirected to the identity provider associated with the organization
* Once the user is authenticated in the identity provider and redirected back to the organization realm, the regular first browser login flow is executed where, at the end, the federated user is added as a member of an organization.

The following constraints will be enforced when onboarding organization members from an identity provider:

* If the email from the identity provider (e.g.: as a result of calling the user info endpoint) does not match the organization domain, the flow will stop
* Using a broker not associated with an organization that matches the email domain of the user will stop the flow
* The authentication flow defaults to the default steps set to the browser flow. For instance, ask for a username and password
* There is no way to choose identity providers when providing the username on the login page
* Linking accounts (when there is an account already with the same email) should work just like if using a regular identity provider (not bound to an organization)
* Only a single identity provider is supported

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #28273

* Allows onboarding organization members by leveraging the identity provider set to an organization
* Changes the `LoginFormsProvider` to allow customizing the attributes available to templates using a general purpose "mapper" rather than a specific method to enable/disable a capability in login pages
* Both browser and brokering flows will change to include additional steps to handle the organization flow when login in and federating users. For now, only new realms will have these flows updated.
* By using an authenticator we can in the future have specific settings at the organization level to configure how the steps should run. I can also think about having an organization specific authentication flow but that is something we can evaluate in the future.
* As we are only enabling a single identity provider by an organization, the redirect to the IdP is automatic. Once we support multiple IdPs to an organization we can think about introducing additional steps to allow users to select which IdP they want to use. Perhaps something similar to https://github.com/sventorben/keycloak-home-idp-discovery.
* To make it easier to obtain the identity provider associated with an org, a new `OrganizationModel.getIdentityProvider` method was added as it removes boilerplate code to resolve the identity provider for a given organization
* A new method `OrganizationProvider.isEnabled` was added to make it easier to check whether a realm supports organizations and if it is enabled to a realm. By doing that, and relying on realm attributes, we can easily check whether a realm should consider processing authorization requests based on its organizations.

### CODE REVIEW

**Review by jonkoops — DISMISSED** [review]

**Review by vramik — COMMENTED** [review]
Thank you @pedroigor for the PR, I went over it briefly and overall it seems good to me. 

As I'm not very familiar with authenticators part, I'd welcome if someone more experienced in the area would take a look to speed up the review and merging. 

I've added few comments/questions, please take a look. IMO we can create follow-up issues for those to not block the PR from merging.

File: model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProviderFactory.java
Comment (pedroigor) [reply model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProviderFactory.java:64]: Yeah, it is intentional. As mentioned in the description, flows are updated only for new realms. 

Reverting authentication flows is something related to your comment on removing the realm attribute. I'm not sure what is the best way to handle this. That is why I'm proposing a switch to enable/disable organization.

wdyt?

File: model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java
Comment (pedroigor) [reply model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java:51]: Yeah, we need that. But this is related to how we are going to introduce organization-related changes to authentication flows.

I think I'll hold using an attribute for checking whether a realm has organizations and just on checking if there are organizations instead. Would that work for you for now?

File: services/src/main/java/org/keycloak/organization/admin/resource/OrganizationIdentityProviderResource.java
Comment (pedroigor) [reply services/src/main/java/org/keycloak/organization/admin/resource/OrganizationIdentityProviderResource.java:139]: Yeah, but I think we shouldn't for organizations. It is actually very confusing how we manage identity providers. There are also expectations from the `IdentityProviderResource` that the `internalId` can be different and for now, keeping the alias unchanged, works better I would say.

File: model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProviderFactory.java
Comment (pedroigor) [reply model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProviderFactory.java]: For now, let's only update the flow for new realms. Created https://github.com/keycloak/keycloak/issues/28804.

File: services/src/main/java/org/keycloak/organization/admin/resource/OrganizationIdentityProviderResource.java
Comment (vramik) [reply services/src/main/java/org/keycloak/organization/admin/resource/OrganizationIdentityProviderResource.java:139]: Should we then add the check that `internalId` is null for the provided `rep` or it matches the `idp` linked to the org?

Comment (pedroigor) [reply services/src/main/java/org/keycloak/organization/admin/resource/OrganizationIdentityProviderResource.java:139]: TBH, I have no clue because this is kinda confusing in the `IdentityProviderResource`.

I'm more thinking of enforcing here that you can't change the alias we always use the alias to lookup/bind the broker for an org.

If that works for you, let's keep that way and perhaps review this before 26.

**Review by jonkoops — DISMISSED** [review]

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationIdentityProviderTest.java
Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationIdentityProviderTest.java:120]: One expects to delete and the other should return a not found given that the resource was deleted already.

Comment (sguilhen) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationIdentityProviderTest.java:120]: Ah, yeah, didn't notice the different status. Sorry for the noise.

**Review by sguilhen — APPROVED** [review]

File: services/src/main/java/org/keycloak/organization/authentication/authenticators/broker/IdpOrganizaitonAuthenticatorFactory.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/organization/authentication/authenticators/broker/IdpOrganizaitonAuthenticatorFactory.java:34]: Typo in the name of the name of this class.

File: services/src/main/java/org/keycloak/organization/authentication/authenticators/browser/OrganizationAuthenticatorFactory.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/organization/authentication/authenticators/browser/OrganizationAuthenticatorFactory.java:41]: I wonder that this should probably return something different than just `Organization` ? How about something like `Organization Identity Provider Redirector` ?

Also method `getHelpText()` should be likely overriden. Otherwise, it would be same as `IdentityProviderAuthenticatorFactory.getHelpText()` , which is not so ideal...

File: services/src/main/java/org/keycloak/organization/authentication/authenticators/broker/IdpOrganizaitonAuthenticatorFactory.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/organization/authentication/authenticators/broker/IdpOrganizaitonAuthenticatorFactory.java:80]: Should this be a bit different value with some additional details?

File: services/src/main/java/org/keycloak/organization/authentication/authenticators/broker/IdpOrganizaitonAuthenticatorFactory.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/organization/authentication/authenticators/broker/IdpOrganizaitonAuthenticatorFactory.java:85]: Should this be a bit different value with some additional details?

**Review by mposolda — COMMENTED** [review]
@pedroigor Nice! Added few minor comments inline. Reviewed especially the parts related to authentication (as that is what @vramik mentioned that he did not reviewed too deeply).

File: services/src/main/java/org/keycloak/organization/authentication/authenticators/browser/OrganizationAuthenticatorFactory.java
Comment (pedroigor) [reply services/src/main/java/org/keycloak/organization/authentication/authenticators/browser/OrganizationAuthenticatorFactory.java:41]: Changed to your suggestion. I was trying to keep names small for the sake of a less verbose UI.

File: services/src/main/java/org/keycloak/organization/authentication/authenticators/broker/IdpOrganizaitonAuthenticatorFactory.java
Comment (pedroigor) [reply services/src/main/java/org/keycloak/organization/authentication/authenticators/broker/IdpOrganizaitonAuthenticatorFactory.java]: Changed to your suggestion. I was trying to keep names small for the sake of a less verbose UI.

**Review by mposolda — APPROVED** [review]

**Review by vramik — APPROVED** [review]

File: services/src/main/java/org/keycloak/organization/admin/resource/OrganizationIdentityProviderResource.java
Comment (vramik) [reply services/src/main/java/org/keycloak/organization/admin/resource/OrganizationIdentityProviderResource.java:139]: That works, I've created https://github.com/keycloak/keycloak/issues/28833 to track it and I'll send PR once this one is merged.

**Pull request merged**
